### DISCUSSION TITLE
Best Practices for Testing LIT Web Components in React

### DISCUSSION THREAD

**Discussion by teifler** [lit/lit#4600] [Q&A]
Hello, community!

We've been developing web components and have successfully tested them in isolation using Web Test Runner. 

Now, we're integrating these web components into a separate React project and are encountering challenges when attempting to test their functionality with Jest. Specifically, we're finding it difficult to access and test elements within the Shadow DOM, as Jest's testing utilities primarily allow interaction with the light DOM.

Could you provide guidance or recommend best practices for testing the functionality of web components (that utilize Shadow DOM) within a React application using Jest? Are there specific configurations, libraries, or techniques that facilitate testing in this scenario?

Any insights or examples would be greatly appreciated!

**Comment by augustjk** [comment]
For future readers, this was asked and answered in the Lit Discord server: https://discord.com/channels/1012791295170859069/1222114682752471050/1222114682752471050

> you may want to try https://www.npmjs.com/package/shadow-dom-testing-library
alternatively, playwright's test runner/framework transparently pierces shadow DOM. none of your test code needs to even be aware of it. it just works
but i suspect you're stuck with jest, so the former is probably your best bet
